diff --git a/package.json b/package.json
index a834b47b..b1306eed 100644
--- a/package.json
+++ b/package.json
@@ -68,6 +68,7 @@
     "flowbite": "^1.8.1",
     "flowbite-react": "^0.6.4",
     "framer-motion": "^10.16.4",
+    "ip-address": "^10.1.0",
     "ip-cidr": "^3.1.0",
     "js-cookie": "^3.0.5",
     "lodash": "^4.17.21",
diff --git a/src/app/(dashboard)/dns/nameservers/page.tsx b/src/app/(dashboard)/dns/nameservers/page.tsx
index 7e66f2b6..b287ea09 100644
--- a/src/app/(dashboard)/dns/nameservers/page.tsx
+++ b/src/app/(dashboard)/dns/nameservers/page.tsx
@@ -7,7 +7,7 @@ import SkeletonTable from "@components/skeletons/SkeletonTable";
 import { RestrictedAccess } from "@components/ui/RestrictedAccess";
 import { usePortalElement } from "@hooks/usePortalElement";
 import useFetchApi from "@utils/api";
-import { ExternalLinkIcon, ServerIcon } from "lucide-react";
+import { ExternalLinkIcon } from "lucide-react";
 import React, { lazy, Suspense } from "react";
 import DNSIcon from "@/assets/icons/DNSIcon";
 import { usePermissions } from "@/contexts/PermissionsProvider";
@@ -15,7 +15,7 @@ import { NameserverGroup } from "@/interfaces/Nameserver";
 import PageContainer from "@/layouts/PageContainer";
 
 const NameserverGroupTable = lazy(
-  () => import("@/modules/dns-nameservers/table/NameserverGroupTable"),
+  () => import("@/modules/dns/nameservers/table/NameserverGroupTable"),
 );
 
 export default function NameServers() {
@@ -40,7 +40,7 @@ export default function NameServers() {
             href={"/dns/nameservers"}
             label={"Nameservers"}
             active
-            icon={<ServerIcon size={13} />}
+            icon={<DNSIcon size={13} />}
           />
         </Breadcrumbs>
         <h1 ref={headingRef}>Nameservers</h1>
diff --git a/src/app/(dashboard)/dns/zones/layout.tsx b/src/app/(dashboard)/dns/zones/layout.tsx
new file mode 100644
index 00000000..640fa1fc
--- /dev/null
+++ b/src/app/(dashboard)/dns/zones/layout.tsx
@@ -0,0 +1,8 @@
+import { globalMetaTitle } from "@utils/meta";
+import type { Metadata } from "next";
+import BlankLayout from "@/layouts/BlankLayout";
+
+export const metadata: Metadata = {
+  title: `Zones - DNS - ${globalMetaTitle}`,
+};
+export default BlankLayout;
diff --git a/src/app/(dashboard)/dns/zones/page.tsx b/src/app/(dashboard)/dns/zones/page.tsx
new file mode 100644
index 00000000..0abf50f1
--- /dev/null
+++ b/src/app/(dashboard)/dns/zones/page.tsx
@@ -0,0 +1,70 @@
+"use client";
+
+import Breadcrumbs from "@components/Breadcrumbs";
+import InlineLink from "@components/InlineLink";
+import Paragraph from "@components/Paragraph";
+import SkeletonTable from "@components/skeletons/SkeletonTable";
+import { RestrictedAccess } from "@components/ui/RestrictedAccess";
+import { usePortalElement } from "@hooks/usePortalElement";
+import useFetchApi from "@utils/api";
+import { ExternalLinkIcon } from "lucide-react";
+import React, { lazy, Suspense } from "react";
+import DNSIcon from "@/assets/icons/DNSIcon";
+import { usePermissions } from "@/contexts/PermissionsProvider";
+import { DNS_ZONE_DOCS_LINK, DNSZone } from "@/interfaces/DNS";
+import PageContainer from "@/layouts/PageContainer";
+import { DNSZonesProvider } from "@/modules/dns/zones/DNSZonesProvider";
+import DNSZoneIcon from "@/assets/icons/DNSZoneIcon";
+
+const DNSZonesTable = lazy(
+  () => import("@/modules/dns/zones/table/DNSZonesTable"),
+);
+
+export default function DNSZonePage() {
+  const { permission } = usePermissions();
+
+  const { data: zones, isLoading } = useFetchApi<DNSZone[]>("/dns/zones");
+
+  const { ref: headingRef, portalTarget } =
+    usePortalElement<HTMLHeadingElement>();
+
+  return (
+    <PageContainer>
+      <div className={"p-default py-6"}>
+        <Breadcrumbs>
+          <Breadcrumbs.Item label={"DNS"} icon={<DNSIcon size={13} />} />
+          <Breadcrumbs.Item
+            href={"/dns/zones"}
+            label={"Zones"}
+            active
+            icon={<DNSZoneIcon size={16} />}
+          />
+        </Breadcrumbs>
+        <h1 ref={headingRef}>Zones</h1>
+        <Paragraph>
+          Manage DNS zones to control domain name resolution for your network.
+        </Paragraph>
+        <Paragraph>
+          Learn more about
+          <InlineLink href={DNS_ZONE_DOCS_LINK} target={"_blank"}>
+            DNS Zones
+            <ExternalLinkIcon size={12} />
+          </InlineLink>
+          in our documentation.
+        </Paragraph>
+      </div>
+
+      <RestrictedAccess page={"DNS Zones"} hasAccess={permission?.dns?.read}>
+        <Suspense fallback={<SkeletonTable />}>
+          <DNSZonesProvider>
+            <DNSZonesTable
+              isLoading={isLoading}
+              headingTarget={portalTarget}
+              data={zones}
+            />
+          </DNSZonesProvider>
+        </Suspense>
+      </RestrictedAccess>
+    </PageContainer>
+  );
+}
diff --git a/src/app/(dashboard)/group/page.tsx b/src/app/(dashboard)/group/page.tsx
index df9d0d1e..38118a48 100644
--- a/src/app/(dashboard)/group/page.tsx
+++ b/src/app/(dashboard)/group/page.tsx
@@ -15,6 +15,7 @@ import { useSearchParams } from "next/navigation";
 import React, { useState } from "react";
 import AccessControlIcon from "@/assets/icons/AccessControlIcon";
 import DNSIcon from "@/assets/icons/DNSIcon";
+import DNSZoneIcon from "@/assets/icons/DNSZoneIcon";
 import NetworkRoutesIcon from "@/assets/icons/NetworkRoutesIcon";
 import PeerIcon from "@/assets/icons/PeerIcon";
 import SetupKeysIcon from "@/assets/icons/SetupKeysIcon";
@@ -24,6 +25,7 @@ import { usePermissions } from "@/contexts/PermissionsProvider";
 import RoutesProvider from "@/contexts/RoutesProvider";
 import { Group, GROUP_TOOLTIP_TEXT } from "@/interfaces/Group";
 import PageContainer from "@/layouts/PageContainer";
+import { GroupDNSZonesSection } from "@/modules/groups/details/GroupDNSZonesSection";
 import { GroupNameserversSection } from "@/modules/groups/details/GroupNameserversSection";
 import { GroupNetworkRoutesSection } from "@/modules/groups/details/GroupNetworkRoutesSection";
 import { GroupPeersSection } from "@/modules/groups/details/GroupPeersSection";
@@ -134,7 +136,9 @@ const validAllGroupTabs = [
   "resources",
   "network-routes",
   "nameservers",
+  "zones",
 ];
+
 const validOtherGroupTabs = ["users", "peers", "setup-keys"];
 
 const GroupOverviewTabs = ({ group }: { group: Group }) => {
@@ -162,6 +166,7 @@ const GroupOverviewTabs = ({ group }: { group: Group }) => {
   const resourcesCount = groupDetails?.resources_count || 0;
   const routesCount = groupDetails?.routes?.length || 0;
   const nameserversCount = groupDetails?.nameservers?.length || 0;
+  const zonesCount = groupDetails?.zones?.length || 0;
   const setupKeysCount = groupDetails?.setupKeys?.length || 0;
 
   return (
@@ -249,6 +254,19 @@ const GroupOverviewTabs = ({ group }: { group: Group }) => {
           {singularize("Nameservers", nameserversCount)}
         </TabsTrigger>
 
+        <TabsTrigger
+          value={"zones"}
+          className={groupDetails === null ? "animate-pulse" : ""}
+        >
+          <DNSZoneIcon
+            size={16}
+            className={
+              "fill-nb-gray-500 group-data-[state=active]/trigger:fill-netbird transition-all"
+            }
+          />
+          {singularize("Zones", zonesCount)}
+        </TabsTrigger>
+
         {group.name !== "All" && (
           <TabsTrigger
             value={"setup-keys"}
@@ -304,6 +322,13 @@ const GroupOverviewTabs = ({ group }: { group: Group }) => {
         />
       </TabsContent>
 
+      <TabsContent value={"zones"} className={"pb-8"}>
+        <GroupDNSZonesSection
+          zones={groupDetails?.zones}
+          isLoading={isLoading}
+        />
+      </TabsContent>
+
       <TabsContent value={"setup-keys"} className={"pb-8"}>
         <GroupSetupKeysSection
           setupKeys={groupDetails?.setupKeys}
diff --git a/src/assets/icons/DNSZoneIcon.tsx b/src/assets/icons/DNSZoneIcon.tsx
new file mode 100644
index 00000000..b08b37b1
--- /dev/null
+++ b/src/assets/icons/DNSZoneIcon.tsx
@@ -0,0 +1,19 @@
+import { iconProperties, IconProps } from "@/assets/icons/IconProperties";
+
+export default function DNSZoneIcon(props: IconProps) {
+  return (
+    <svg
+      width="24"
+      height="24"
+      viewBox="0 0 24 24"
+      xmlns="http://www.w3.org/2000/svg"
+      {...iconProperties(props)}
+    >
+      <path
+        fillRule="evenodd"
+        d="M5 5a2 2 0 0 0-2 2v3a1 1 0 0 0 1 1h16a1 1 0 0 0 1-1V7a2 2 0 0 0-2-2H5Zm9 2a1 1 0 1 0 0 2h.01a1 1 0 1 0 0-2H14Zm3 0a1 1 0 1 0 0 2h.01a1 1 0 1 0 0-2H17ZM3 17v-3a1 1 0 0 1 1-1h16a1 1 0 0 1 1 1v3a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2Zm11-2a1 1 0 1 0 0 2h.01a1 1 0 1 0 0-2H14Zm3 0a1 1 0 1 0 0 2h.01a1 1 0 1 0 0-2H17Z"
+        clipRule="evenodd"
+      />
+    </svg>
+  );
+}
diff --git a/src/components/table/Table.tsx b/src/components/table/Table.tsx
index fc119926..ac0b0cb3 100644
--- a/src/components/table/Table.tsx
+++ b/src/components/table/Table.tsx
@@ -104,7 +104,7 @@ const TableRow = React.forwardRef<
       " transition-colors group/table-row data-[state=selected]:bg-neutral-100 dark:data-[state=selected]:bg-nb-gray-930/70",
       "dark:data-[state=selected]:border-nb-gray-900",
       minimal
-        ? "dark:hover:bg-nb-gray-900/10"
+        ? "dark:hover:bg-nb-gray-910/[15%]"
         : "border-b dark:border-zinc-700/40 dark:hover:bg-nb-gray-940 hover:bg-neutral-100/50",
       className,
     )}
diff --git a/src/components/ui/PeerCountBadge.tsx b/src/components/ui/PeerCountBadge.tsx
index 73b6d256..f8fa4ce5 100644
--- a/src/components/ui/PeerCountBadge.tsx
+++ b/src/components/ui/PeerCountBadge.tsx
@@ -19,11 +19,12 @@ export default function PeerCountBadge({
   className,
 }: Props) {
   const router = useRouter();
-  const { dropdownOptions } = useGroups();
+  const { dropdownOptions, groups } = useGroups();
 
   const currentGroup = useMemo(() => {
-    return dropdownOptions?.find((g) => g.name === group?.name);
-  }, [group, dropdownOptions]);
+    const options = dropdownOptions?.find((g) => g.name === group?.name);
+    return options ?? groups?.find((g) => g.name === group?.name);
+  }, [group, dropdownOptions, groups]);
 
   const peerCount = useMemo(() => {
     let peerCount = currentGroup?.peers_count ?? 0;
diff --git a/src/contexts/DialogProvider.tsx b/src/contexts/DialogProvider.tsx
index 0cf10d7d..096b28f7 100644
--- a/src/contexts/DialogProvider.tsx
+++ b/src/contexts/DialogProvider.tsx
@@ -66,6 +66,8 @@ export default function DialogProvider({ children }: Props) {
           <ModalContent
             maxWidthClass={dialogOptions.maxWidthClass || "max-w-[400px]"}
             showClose={false}
+            onInteractOutside={(e) => e.preventDefault()}
+            onPointerDownOutside={(e) => e.preventDefault()}
           >
             <ModalHeader
               center={dialogOptions.type == "center"}
diff --git a/src/interfaces/DNS.ts b/src/interfaces/DNS.ts
new file mode 100644
index 00000000..58daa965
--- /dev/null
+++ b/src/interfaces/DNS.ts
@@ -0,0 +1,23 @@
+export interface DNSZone {
+  id?: string;
+  name: string;
+  domain: string;
+  enabled: boolean;
+  enable_search_domain: boolean;
+  distribution_groups: string[];
+  records?: DNSRecord[];
+  groups_search?: string;
+}
+
+export interface DNSRecord {
+  id?: string;
+  name: string;
+  type: "A" | "AAAA" | "CNAME";
+  content: string;
+  ttl: number;
+}
+
+export type DNSRecordType = "A" | "AAAA" | "CNAME";
+
+export const DNS_ZONE_DOCS_LINK = "https://docs.netbird.io/manage/dns/zones";
+export const DNS_RECORDS_DOCS_LINK = "https://docs.netbird.io/manage/dns/zones";
diff --git a/src/layouts/Navigation.tsx b/src/layouts/Navigation.tsx
index 9c611a25..224f7a99 100644
--- a/src/layouts/Navigation.tsx
+++ b/src/layouts/Navigation.tsx
@@ -146,6 +146,12 @@ export default function Navigation({
                     href={"/dns/nameservers"}
                     visible={permission.nameservers.read}
                   />
+                  <SidebarItem
+                    label="Zones"
+                    isChild
+                    href={"/dns/zones"}
+                    visible={permission?.dns?.read}
+                  />
                   <SidebarItem
                     label="DNS Settings"
                     isChild
diff --git a/src/modules/dns-nameservers/NameserverModal.tsx b/src/modules/dns/nameservers/NameserverModal.tsx
similarity index 100%
rename from src/modules/dns-nameservers/NameserverModal.tsx
rename to src/modules/dns/nameservers/NameserverModal.tsx
diff --git a/src/modules/dns-nameservers/NameserverTemplateModal.tsx b/src/modules/dns/nameservers/NameserverTemplateModal.tsx
similarity index 98%
rename from src/modules/dns-nameservers/NameserverTemplateModal.tsx
rename to src/modules/dns/nameservers/NameserverTemplateModal.tsx
index dc642168..73541bbc 100644
--- a/src/modules/dns-nameservers/NameserverTemplateModal.tsx
+++ b/src/modules/dns/nameservers/NameserverTemplateModal.tsx
@@ -9,7 +9,7 @@ import GoogleLogo from "@/assets/nameservers/google.svg";
 import Quad9Logo from "@/assets/nameservers/quad9.svg";
 import { Group } from "@/interfaces/Group";
 import { NameserverGroup, NameserverPresets } from "@/interfaces/Nameserver";
-import NameserverModal from "@/modules/dns-nameservers/NameserverModal";
+import NameserverModal from "@/modules/dns/nameservers/NameserverModal";
 
 type Props = {
   children: React.ReactNode;
diff --git a/src/modules/dns-nameservers/table/NameserverActionCell.tsx b/src/modules/dns/nameservers/table/NameserverActionCell.tsx
similarity index 100%
rename from src/modules/dns-nameservers/table/NameserverActionCell.tsx
rename to src/modules/dns/nameservers/table/NameserverActionCell.tsx
diff --git a/src/modules/dns-nameservers/table/NameserverActiveCell.tsx b/src/modules/dns/nameservers/table/NameserverActiveCell.tsx
similarity index 100%
rename from src/modules/dns-nameservers/table/NameserverActiveCell.tsx
rename to src/modules/dns/nameservers/table/NameserverActiveCell.tsx
diff --git a/src/modules/dns-nameservers/table/NameserverDistributionGroupsCell.tsx b/src/modules/dns/nameservers/table/NameserverDistributionGroupsCell.tsx
similarity index 100%
rename from src/modules/dns-nameservers/table/NameserverDistributionGroupsCell.tsx
rename to src/modules/dns/nameservers/table/NameserverDistributionGroupsCell.tsx
diff --git a/src/modules/dns-nameservers/table/NameserverGroupTable.tsx b/src/modules/dns/nameservers/table/NameserverGroupTable.tsx
similarity index 95%
rename from src/modules/dns-nameservers/table/NameserverGroupTable.tsx
rename to src/modules/dns/nameservers/table/NameserverGroupTable.tsx
index b668e8dc..69f81dc7 100644
--- a/src/modules/dns-nameservers/table/NameserverGroupTable.tsx
+++ b/src/modules/dns/nameservers/table/NameserverGroupTable.tsx
@@ -19,14 +19,14 @@ import { usePermissions } from "@/contexts/PermissionsProvider";
 import { useLocalStorage } from "@/hooks/useLocalStorage";
 import { Group } from "@/interfaces/Group";
 import { NameserverGroup } from "@/interfaces/Nameserver";
-import NameserverModal from "@/modules/dns-nameservers/NameserverModal";
-import NameserverTemplateModal from "@/modules/dns-nameservers/NameserverTemplateModal";
-import NameserverActionCell from "@/modules/dns-nameservers/table/NameserverActionCell";
-import NameserverActiveCell from "@/modules/dns-nameservers/table/NameserverActiveCell";
-import NameserverDistributionGroupsCell from "@/modules/dns-nameservers/table/NameserverDistributionGroupsCell";
-import NameserverMatchDomainsCell from "@/modules/dns-nameservers/table/NameserverMatchDomainsCell";
-import NameserverNameCell from "@/modules/dns-nameservers/table/NameserverNameCell";
-import NameserverNameserversCell from "@/modules/dns-nameservers/table/NameserverNameserversCell";
+import NameserverModal from "@/modules/dns/nameservers/NameserverModal";
+import NameserverTemplateModal from "@/modules/dns/nameservers/NameserverTemplateModal";
+import NameserverActionCell from "@/modules/dns/nameservers/table/NameserverActionCell";
+import NameserverActiveCell from "@/modules/dns/nameservers/table/NameserverActiveCell";
+import NameserverDistributionGroupsCell from "@/modules/dns/nameservers/table/NameserverDistributionGroupsCell";
+import NameserverMatchDomainsCell from "@/modules/dns/nameservers/table/NameserverMatchDomainsCell";
+import NameserverNameCell from "@/modules/dns/nameservers/table/NameserverNameCell";
+import NameserverNameserversCell from "@/modules/dns/nameservers/table/NameserverNameserversCell";
 
 export const NameserverGroupTableColumns: ColumnDef<NameserverGroup>[] = [
   {
diff --git a/src/modules/dns-nameservers/table/NameserverMatchDomainsCell.tsx b/src/modules/dns/nameservers/table/NameserverMatchDomainsCell.tsx
similarity index 100%
rename from src/modules/dns-nameservers/table/NameserverMatchDomainsCell.tsx
rename to src/modules/dns/nameservers/table/NameserverMatchDomainsCell.tsx
diff --git a/src/modules/dns-nameservers/table/NameserverNameCell.tsx b/src/modules/dns/nameservers/table/NameserverNameCell.tsx
similarity index 100%
rename from src/modules/dns-nameservers/table/NameserverNameCell.tsx
rename to src/modules/dns/nameservers/table/NameserverNameCell.tsx
diff --git a/src/modules/dns-nameservers/table/NameserverNameserversCell.tsx b/src/modules/dns/nameservers/table/NameserverNameserversCell.tsx
similarity index 100%
rename from src/modules/dns-nameservers/table/NameserverNameserversCell.tsx
rename to src/modules/dns/nameservers/table/NameserverNameserversCell.tsx
diff --git a/src/modules/dns/zones/DNSRecordModal.tsx b/src/modules/dns/zones/DNSRecordModal.tsx
new file mode 100644
index 00000000..7e59b2b1
--- /dev/null
+++ b/src/modules/dns/zones/DNSRecordModal.tsx
@@ -0,0 +1,359 @@
+import Button from "@components/Button";
+import HelpText from "@components/HelpText";
+import InlineLink from "@components/InlineLink";
+import { Input } from "@components/Input";
+import { Label } from "@components/Label";
+import {
+  Modal,
+  ModalClose,
+  ModalContent,
+  ModalFooter,
+  ModalTrigger,
+} from "@components/modal/Modal";
+import ModalHeader from "@components/modal/ModalHeader";
+import Paragraph from "@components/Paragraph";
+import {
+  Select,
+  SelectContent,
+  SelectItem,
+  SelectTrigger,
+  SelectValue,
+} from "@components/Select";
+import Separator from "@components/Separator";
+import { validator } from "@utils/helpers";
+import { Address4, Address6 } from "ip-address";
+import { ClockIcon, ExternalLinkIcon, GlobeIcon } from "lucide-react";
+import React, { useMemo, useState } from "react";
+import {
+  DNS_RECORDS_DOCS_LINK,
+  DNSRecord,
+  DNSRecordType,
+  DNSZone,
+} from "@/interfaces/DNS";
+import { useDNSZones } from "@/modules/dns/zones/DNSZonesProvider";
+
+type Props = {
+  children?: React.ReactNode;
+  open: boolean;
+  onOpenChange: (open: boolean) => void;
+  zone: DNSZone;
+  record?: DNSRecord;
+};
+
+export default function DNSRecordModal({
+  children,
+  open,
+  onOpenChange,
+  zone,
+  record,
+}: Readonly<Props>) {
+  return (
+    <Modal open={open} onOpenChange={onOpenChange} key={open ? 1 : 0}>
+      {children && <ModalTrigger asChild>{children}</ModalTrigger>}
+      {open && (
+        <DNSRecordModalContent
+          onSuccess={() => onOpenChange(false)}
+          onSuccessAdded={() => {
+            setTimeout(() => {
+              const row = document.querySelector<HTMLElement>(
+                `[data-row-id="${zone.id}"]`,
+              );
+              if (row?.getAttribute("data-accordion") === "closed") {
+                row?.click();
+              }
+              row?.scrollIntoView({ behavior: "smooth" });
+            }, 200);
+            onOpenChange(false);
+          }}
+          zone={zone}
+          record={record}
+        />
+      )}
+    </Modal>
+  );
+}
+
+type ModalProps = {
+  onSuccess?: () => void;
+  onSuccessAdded?: () => void;
+  zone: DNSZone;
+  record?: DNSRecord;
+};
+
+export function DNSRecordModalContent({
+  onSuccess,
+  onSuccessAdded,
+  zone,
+  record,
+}: Readonly<ModalProps>) {
+  const { addRecord, updateRecord } = useDNSZones();
+
+  const getInitialDomain = () => {
+    if (!record) return "";
+    if (record.name === zone.domain) return "";
+    return record.name.replace(`.${zone.domain}`, "");
+  };
+
+  const [domain, setDomain] = useState(record?.name ? getInitialDomain() : "");
+  const [ttl, setTtl] = useState(record ? record.ttl.toString() : "300");
+  const [type, setType] = useState<DNSRecordType>(record?.type ?? "A");
+  const [recordValue, setRecordValue] = useState(record?.content ?? "");
+
+  const domainError = useMemo(() => {
+    if (domain == "") return "";
+    const valid = validator.isValidDomain(domain, {
+      allowWildcard: false,
+      allowOnlyTld: true,
+    });
+    if (!valid) {
+      return "Please enter a valid domain, e.g. example.com or intra.example.com";
+    }
+  }, [domain]);
+
+  const ipv4Error = useMemo(() => {
+    if (recordValue === "" || type !== "A") return "";
+    const valid = Address4.isValid(recordValue);
+    if (!valid) {
+      return "Please enter a valid IPv4 address, e.g. 192.168.1.1";
+    }
+  }, [recordValue, type]);
+
+  const ipv6Error = useMemo(() => {
+    if (recordValue === "" || type !== "AAAA") return "";
+    const valid = Address6.isValid(recordValue);
+    if (!valid) {
+      return "Please enter a valid IPv6 address, e.g. 2001:0db8:85a3::8a2e:0370:7334";
+    }
+  }, [recordValue, type]);
+
+  const cnameError = useMemo(() => {
+    if (recordValue === "" || type !== "CNAME") return "";
+    const valid = validator.isValidDomain(recordValue, {
+      allowWildcard: false,
+      allowOnlyTld: false,
+    });
+    if (!valid) {
+      return "Please enter a valid domain, e.g. example.com or server.example.com";
+    }
+  }, [recordValue, type]);
+
+  const handleAddRecord = async () => {
+    const name = domain !== "" ? `${domain}.${zone.domain}` : zone.domain;
+
+    if (record) {
+      updateRecord(zone, {
+        id: record.id,
+        name,
+        type,
+        content: recordValue,
+        ttl: parseInt(ttl),
+      }).then(onSuccess);
+    } else {
+      addRecord(zone, {
+        name,
+        type,
+        content: recordValue,
+        ttl: parseInt(ttl),
+      }).then(onSuccessAdded);
+    }
+  };
+
+  const canUpdateOrCreate =
+    !cnameError &&
+    !ipv6Error &&
+    !ipv4Error &&
+    !domainError &&
+    recordValue !== "";
+
+  return (
+    <ModalContent maxWidthClass={"max-w-xl"}>
+      <ModalHeader
+        title={record ? "Update DNS Record" : "Add DNS Record"}
+        description={
+          record
+            ? `Update record of '${zone.domain}' zone`
+            : `Add new record to the '${zone.domain}' zone`
+        }
+        icon={<GlobeIcon size={16} />}
+      />
+      <Separator />
+      <div className={"px-8 py-6 flex flex-col gap-6"}>
+        <div className={"flex items-center justify-between gap-10"}>
+          <div>
+            <Label>Record Type</Label>
+            <HelpText className={"max-w-sm"}>
+              Select the type of record you want to add
+            </HelpText>
+          </div>
+          <div className={"min-w-[130px]"}>
+            <Select
+              value={type}
+              onValueChange={(v) => {
+                setType(v as DNSRecordType);
+                setRecordValue("");
+              }}
+            >
+              <SelectTrigger
+                className="w-full pl-4"
+                data-cy={"dns-record-type-select"}
+              >
+                <SelectValue placeholder="Select type..." />
+              </SelectTrigger>
+              <SelectContent>
+                <SelectItem value="A">A</SelectItem>
+                <SelectItem value="AAAA">AAAA</SelectItem>
+                <SelectItem value="CNAME">CNAME</SelectItem>
+              </SelectContent>
+            </Select>
+          </div>
+        </div>
+        <div className={"w-full mb-3"}>
+          <Label>Hostname</Label>
+          <HelpText>
+            Enter a subdomain or leave empty to use the primary domain.
+          </HelpText>
+          <div className={"flex w-full"}>
+            <Input
+              autoFocus={true}
+              placeholder={"Subdomain (leave empty for primary domain)"}
+              errorTooltip={true}
+              errorTooltipPosition={"bottom"}
+              error={domainError}
+              value={domain}
+              className={"rounded-r-none"}
+              maxWidthClass={"w-full"}
+              onChange={(e) => setDomain(e.target.value)}
+            />
+            <div
+              className={
+                "bg-nb-gray-900 rounded-r-md border text-nb-gray-300 border-l-0 text-sm border-nb-gray-700 flex items-center justify-center whitespace-nowrap px-4 opacity-80"
+              }
+            >
+              .{zone.domain}
+            </div>
+          </div>
+        </div>
+
+        <div className={"flex gap-4 items-start mb-3"}>
+          {type === "A" && (
+            <div className={"flex-1"}>
+              <Label>IPv4 Address</Label>
+              <Input
+                className={"mt-1.5 font-mono text-[0.82rem]"}
+                placeholder={"192.168.1.1"}
+                errorTooltip={false}
+                errorTooltipPosition={"top"}
+                error={ipv4Error}
+                value={recordValue}
+                maxWidthClass={"w-full"}
+                onChange={(e) => setRecordValue(e.target.value)}
+              />
+            </div>
+          )}
+
+          {type === "AAAA" && (
+            <div className={"flex-1"}>
+              <Label>IPv6 Address</Label>
+              <Input
+                className={"mt-1.5 font-mono text-[0.82rem]"}
+                placeholder={"2001:0db8:85a3::8a2e:0370:7334"}
+                errorTooltip={false}
+                errorTooltipPosition={"top"}
+                error={ipv6Error}
+                value={recordValue}
+                maxWidthClass={"w-full"}
+                onChange={(e) => setRecordValue(e.target.value)}
+              />
+            </div>
+          )}
+
+          {type === "CNAME" && (
+            <div className={"flex-1"}>
+              <Label>Target Domain</Label>
+              <Input
+                className={"mt-1.5"}
+                placeholder={"e.g., example.com or intra.example.com"}
+                errorTooltip={false}
+                errorTooltipPosition={"top"}
+                error={cnameError}
+                value={recordValue}
+                maxWidthClass={"w-full"}
+                onChange={(e) => setRecordValue(e.target.value)}
+              />
+            </div>
+          )}
+
+          <div className={"min-w-[200px]"}>
+            <Label>TTL (Time to Live)</Label>
+            <div className={"mt-2.5"}>
+              <Select value={ttl} onValueChange={(v) => setTtl(v)}>
+                <SelectTrigger
+                  className="w-full"
+                  data-cy={"dns-record-ttl-select"}
+                >
+                  <div className={"flex items-center gap-2"}>
+                    <ClockIcon size={14} className={"text-nb-gray-300"} />
+                    <SelectValue placeholder="Select TTL..." />
+                  </div>
+                </SelectTrigger>
+                <SelectContent>
+                  <SelectItem value="60">{getTTLLabel(60)}</SelectItem>
+                  <SelectItem value="120">{getTTLLabel(120)}</SelectItem>
+                  <SelectItem value="300">{getTTLLabel(300)}</SelectItem>
+                  <SelectItem value="600">{getTTLLabel(600)}</SelectItem>
+                  <SelectItem value="900">{getTTLLabel(900)}</SelectItem>
+                  <SelectItem value="1800">{getTTLLabel(1800)}</SelectItem>
+                  <SelectItem value="3600">{getTTLLabel(3600)}</SelectItem>
+                  <SelectItem value="7200">{getTTLLabel(7200)}</SelectItem>
+                  <SelectItem value="43200">{getTTLLabel(43200)}</SelectItem>
+                  <SelectItem value="86400">{getTTLLabel(86400)}</SelectItem>
+                </SelectContent>
+              </Select>
+            </div>
+          </div>
+        </div>
+      </div>
+
+      <ModalFooter className={"items-center"}>
+        <div className={"w-full"}>
+          <Paragraph className={"text-sm mt-auto"}>
+            Learn more about
+            <InlineLink href={DNS_RECORDS_DOCS_LINK} target={"_blank"}>
+              DNS Records
+              <ExternalLinkIcon size={12} />
+            </InlineLink>
+          </Paragraph>
+        </div>
+
+        <div className={"flex gap-3 w-full justify-end"}>
+          <>
+            <ModalClose asChild={true}>
+              <Button variant={"secondary"}>Cancel</Button>
+            </ModalClose>
+            <Button
+              variant={"primary"}
+              onClick={handleAddRecord}
+              disabled={!canUpdateOrCreate}
+            >
+              {record ? "Save Changes" : "Add Record"}
+            </Button>
+          </>
+        </div>
+      </ModalFooter>
+    </ModalContent>
+  );
+}
+
+export const getTTLLabel = (seconds: number): string => {
+  if (seconds < 60) return `${seconds} Sec.`;
+  if (seconds < 3600) {
+    const minutes = seconds / 60;
+    return minutes === 1 ? "1 Min." : `${minutes} Min.`;
+  }
+  if (seconds < 86400) {
+    const hours = seconds / 3600;
+    return hours === 1 ? "1 Hour" : `${hours} Hours`;
+  }
+  const days = seconds / 86400;
+  return days === 1 ? "1 Day" : `${days} Days`;
+};
diff --git a/src/modules/dns/zones/DNSZoneModal.tsx b/src/modules/dns/zones/DNSZoneModal.tsx
new file mode 100644
index 00000000..4767a6d6
--- /dev/null
+++ b/src/modules/dns/zones/DNSZoneModal.tsx
@@ -0,0 +1,225 @@
+import Button from "@components/Button";
+import FancyToggleSwitch from "@components/FancyToggleSwitch";
+import HelpText from "@components/HelpText";
+import InlineLink from "@components/InlineLink";
+import { Input } from "@components/Input";
+import { Label } from "@components/Label";
+import {
+  Modal,
+  ModalClose,
+  ModalContent,
+  ModalFooter,
+  ModalTrigger,
+} from "@components/modal/Modal";
+import ModalHeader from "@components/modal/ModalHeader";
+import Paragraph from "@components/Paragraph";
+import { PeerGroupSelector } from "@components/PeerGroupSelector";
+import Separator from "@components/Separator";
+import { validator } from "@utils/helpers";
+import { ExternalLinkIcon, Power, ScanSearch } from "lucide-react";
+import React, { useMemo, useState } from "react";
+import { DNS_ZONE_DOCS_LINK, DNSZone } from "@/interfaces/DNS";
+import { useDNSZones } from "@/modules/dns/zones/DNSZonesProvider";
+import useGroupHelper from "@/modules/groups/useGroupHelper";
+import { Group } from "@/interfaces/Group";
+import DNSZoneIcon from "@/assets/icons/DNSZoneIcon";
+
+type Props = {
+  children?: React.ReactNode;
+  open: boolean;
+  onOpenChange: (open: boolean) => void;
+  onSuccess?: (zone: DNSZone) => void;
+  onSuccessAdded?: (zone: DNSZone) => void;
+  initialDistributionGroups?: Group[];
+  zone?: DNSZone;
+};
+
+export default function DNSZoneModal({
+  children,
+  open,
+  onOpenChange,
+  onSuccess,
+  onSuccessAdded,
+  initialDistributionGroups,
+  zone,
+}: Readonly<Props>) {
+  return (
+    <Modal open={open} onOpenChange={onOpenChange} key={open ? 1 : 0}>
+      {children && <ModalTrigger asChild>{children}</ModalTrigger>}
+      {open && (
+        <DNSZoneModalContent
+          onSuccess={(z) => {
+            onOpenChange(false);
+            onSuccess?.(z);
+          }}
+          onSuccessAdded={(z) => {
+            onOpenChange(false);
+            onSuccessAdded?.(z);
+          }}
+          zone={zone}
+          initialDistributionGroups={initialDistributionGroups}
+        />
+      )}
+    </Modal>
+  );
+}
+
+type ModalProps = {
+  onSuccess?: (zone: DNSZone) => void;
+  onSuccessAdded?: (zone: DNSZone) => void;
+  initialDistributionGroups?: Group[];
+  zone?: DNSZone;
+};
+
+export function DNSZoneModalContent({
+  onSuccess,
+  onSuccessAdded,
+  zone,
+  initialDistributionGroups,
+}: Readonly<ModalProps>) {
+  const { createZone, updateZone } = useDNSZones();
+  const [domain, setDomain] = useState(zone?.domain ?? "");
+  const [enabled, setEnabled] = useState<boolean>(zone?.enabled ?? true);
+  const [searchDomainsEnabled, setSearchDomainsEnabled] = useState(
+    zone?.enable_search_domain ?? false,
+  );
+  const [groups, setGroups, { save: saveGroups }] = useGroupHelper({
+    initial: initialDistributionGroups ?? zone?.distribution_groups ?? [],
+  });
+
+  const domainError = useMemo(() => {
+    if (domain == "") return "";
+    const valid = validator.isValidDomain(domain, {
+      allowWildcard: false,
+      allowOnlyTld: false,
+    });
+    if (!valid) {
+      return "Please enter a valid domain, e.g. company.internal or intra.example.com";
+    }
+  }, [domain]);
+
+  const handleOnSubmit = async () => {
+    return saveGroups().then((distributionGroups) => {
+      const groupIds = distributionGroups.map((group) => group.id as string);
+
+      if (zone) {
+        updateZone({
+          id: zone.id,
+          domain,
+          name: domain,
+          distribution_groups: groupIds,
+          enabled,
+          enable_search_domain: searchDomainsEnabled,
+        } as DNSZone).then(onSuccess);
+      } else {
+        createZone({
+          domain,
+          name: domain,
+          distribution_groups: groupIds,
+          enabled,
+          enable_search_domain: searchDomainsEnabled,
+        } as DNSZone).then(onSuccessAdded);
+      }
+    });
+  };
+
+  const canUpdateOrCreate = !domainError && groups?.length > 0 && domain !== "";
+
+  return (
+    <ModalContent maxWidthClass={"max-w-2xl"}>
+      <ModalHeader
+        icon={<DNSZoneIcon size={20} className={"fill-netbird"} />}
+        title={zone ? "Update DNS Zone" : "Add DNS Zone"}
+        description={
+          "Use a zone to control domain name resolution for your network."
+        }
+        color={"netbird"}
+      />
+
+      <Separator />
+
+      <div className={"px-8 pt-6 pb-7 flex-col flex gap-6"}>
+        <div>
+          <Label>Domain</Label>
+          <HelpText>
+            Enter a domain for this zone (e.g., company.internal,
+            intra.example.com)
+          </HelpText>
+          <Input
+            disabled={!!zone}
+            readOnly={!!zone}
+            placeholder={"e.g., company.internal"}
+            errorTooltip={false}
+            errorTooltipPosition={"top"}
+            error={domainError}
+            value={domain}
+            onChange={(e) => setDomain(e.target.value)}
+          />
+        </div>
+        <div className={"mb-2"}>
+          <Label>Distribution Groups</Label>
+          <HelpText>
+            Advertise this zone and its records to peers that belong to the
+            following groups
+          </HelpText>
+          <PeerGroupSelector
+            onChange={setGroups}
+            values={groups}
+            showResources={false}
+            showResourceCounter={false}
+          />
+        </div>
+
+        <FancyToggleSwitch
+          value={searchDomainsEnabled}
+          onChange={setSearchDomainsEnabled}
+          label={
+            <>
+              <ScanSearch size={15} />
+              Enable Search Domains
+            </>
+          }
+          helpText={
+            "E.g., 'server.company.internal' will be accessible with 'server'"
+          }
+        />
+
+        <FancyToggleSwitch
+          value={enabled}
+          onChange={setEnabled}
+          label={
+            <>
+              <Power size={15} />
+              Enable DNS Zone
+            </>
+          }
+          helpText={"Use this switch to enable or disable the dns zone."}
+        />
+      </div>
+
+      <ModalFooter className={"items-center"}>
+        <div className={"w-full"}>
+          <Paragraph className={"text-sm mt-auto"}>
+            Learn more about
+            <InlineLink href={DNS_ZONE_DOCS_LINK} target={"_blank"}>
+              DNS Zones
+              <ExternalLinkIcon size={12} />
+            </InlineLink>
+          </Paragraph>
+        </div>
+        <div className={"flex gap-3 w-full justify-end"}>
+          <ModalClose asChild={true}>
+            <Button variant={"secondary"}>Cancel</Button>
+          </ModalClose>
+          <Button
+            variant={"primary"}
+            onClick={handleOnSubmit}
+            disabled={!canUpdateOrCreate}
+          >
+            {zone ? "Save Changes" : "Add Zone"}
+          </Button>
+        </div>
+      </ModalFooter>
+    </ModalContent>
+  );
+}
diff --git a/src/modules/dns/zones/DNSZonesProvider.tsx b/src/modules/dns/zones/DNSZonesProvider.tsx
new file mode 100644
index 00000000..c3b3e4d3
--- /dev/null
+++ b/src/modules/dns/zones/DNSZonesProvider.tsx
@@ -0,0 +1,264 @@
+import { notify } from "@components/Notification";
+import { useApiCall } from "@utils/api";
+import * as React from "react";
+import { useState } from "react";
+import { useSWRConfig } from "swr";
+import { useDialog } from "@/contexts/DialogProvider";
+import { DNSRecord, DNSZone } from "@/interfaces/DNS";
+import { Group } from "@/interfaces/Group";
+import DNSRecordModal from "@/modules/dns/zones/DNSRecordModal";
+import DNSZoneModal from "@/modules/dns/zones/DNSZoneModal";
+
+type Props = {
+  children?: React.ReactNode;
+};
+
+const DNSZonesContext = React.createContext(
+  {} as {
+    createZone: (zone: DNSZone) => Promise<DNSZone>;
+    updateZone: (zone: DNSZone) => Promise<DNSZone>;
+    deleteZone: (zone: DNSZone) => Promise<DNSZone>;
+    openZoneModal: (
+      zone?: DNSZone,
+      initialDistributionGroups?: Group[],
+    ) => void;
+    openRecordModal: (zone: DNSZone, record?: DNSRecord) => void;
+    addRecord: (zone: DNSZone, record: DNSRecord) => Promise<DNSRecord>;
+    updateRecord: (zone: DNSZone, record: DNSRecord) => Promise<DNSRecord>;
+    deleteRecord: (zone: DNSZone, record: DNSRecord) => Promise<DNSRecord>;
+    askForRecord: (zone: DNSZone) => void;
+  },
+);
+
+export const DNSZonesProvider = ({ children }: Props) => {
+  const { mutate } = useSWRConfig();
+  const zoneRequest = useApiCall<DNSZone>("/dns/zones", true);
+  const recordRequest = useApiCall<DNSRecord>("/dns/zones", true);
+  const [dnsModal, setDnsModal] = useState(false);
+  const [recordModal, setRecordModal] = useState(false);
+  const [currentZone, setCurrentZone] = useState<DNSZone>();
+  const [currentRecord, setCurrentRecord] = useState<DNSRecord>();
+  const [initialDistributionGroups, setInitialDistributionGroups] =
+    useState<Group[]>();
+  const { confirm } = useDialog();
+
+  const createZone = async (zone: DNSZone): Promise<DNSZone> => {
+    const promise = zoneRequest.post(zone).then((zone) => {
+      mutate("/dns/zones");
+      return Promise.resolve(zone);
+    });
+
+    notify({
+      title: `DNS Zone '${zone.domain}'`,
+      description: `DNS Zone was added successfully.`,
+      promise: promise,
+      loadingMessage: "Adding DNS Zone...",
+    });
+
+    return promise;
+  };
+
+  const updateZone = async (zone: DNSZone): Promise<DNSZone> => {
+    if (!zone?.id) return Promise.reject("Can not update DNS Zone without ID");
+    const promise = zoneRequest.put(zone, `/${zone.id}`).then((zone) => {
+      mutate("/dns/zones");
+      return Promise.resolve(zone);
+    });
+
+    notify({
+      title: `DNS Zone '${zone.domain}'`,
+      description: `DNS Zone was updated successfully.`,
+      promise: promise,
+      loadingMessage: "Updating DNS Zone...",
+    });
+
+    return promise;
+  };
+
+  const deleteZone = async (zone: DNSZone): Promise<DNSZone> => {
+    if (!zone?.id) return Promise.reject("Can not delete DNS Zone without ID");
+
+    const choice = await confirm({
+      title: `Delete zone '${zone.domain}'?`,
+      description:
+        "Are you sure you want to delete this zone? This action cannot be undone.",
+      confirmText: "Delete",
+      cancelText: "Cancel",
+      type: "danger",
+      maxWidthClass: "max-w-md",
+    });
+    if (!choice) return Promise.resolve(zone);
+
+    const promise = zoneRequest.del({}, `/${zone.id}`).then((zone) => {
+      mutate("/dns/zones");
+      return Promise.resolve(zone);
+    });
+
+    notify({
+      title: `DNS Zone '${zone.domain}'`,
+      description: `DNS Zone was deleted successfully.`,
+      promise: promise,
+      loadingMessage: "Deleting DNS Zone...",
+    });
+
+    return promise;
+  };
+
+  const addRecord = async (
+    zone: DNSZone,
+    record: DNSRecord,
+  ): Promise<DNSRecord> => {
+    if (!zone?.id)
+      return Promise.reject("Can not add DNS Record without DNS Zone");
+    const promise = recordRequest
+      .post(record, `/${zone.id}/records`)
+      .then((record) => {
+        mutate("/dns/zones");
+        return Promise.resolve(record);
+      });
+
+    notify({
+      title: `${record.type} Record '${record.name}'`,
+      description: `DNS Record was added successfully.`,
+      promise: promise,
+      loadingMessage: "Adding DNS Record...",
+    });
+
+    return promise;
+  };
+
+  const updateRecord = async (
+    zone: DNSZone,
+    record: DNSRecord,
+  ): Promise<DNSRecord> => {
+    if (!zone?.id)
+      return Promise.reject("Can not update DNS Record without DNS Zone");
+    if (!record?.id)
+      return Promise.reject("Can not update DNS Record without ID");
+    const promise = recordRequest
+      .put(record, `/${zone.id}/records/${record.id}`)
+      .then((record) => {
+        mutate("/dns/zones");
+        return Promise.resolve(record);
+      });
+
+    notify({
+      title: `${record.type} Record '${record.name}'`,
+      description: `DNS Record was updated successfully.`,
+      promise: promise,
+      loadingMessage: "Updating DNS Record...",
+    });
+
+    return promise;
+  };
+
+  const deleteRecord = async (
+    zone: DNSZone,
+    record: DNSRecord,
+  ): Promise<DNSRecord> => {
+    if (!zone?.id)
+      return Promise.reject("Can not delete DNS Record without DNS Zone");
+    if (!record?.id)
+      return Promise.reject("Can not delete DNS Record without ID");
+
+    const choice = await confirm({
+      title: `Delete record '${record.name}'?`,
+      description:
+        "Are you sure you want to delete this record? This action cannot be undone.",
+      confirmText: "Delete",
+      cancelText: "Cancel",
+      type: "danger",
+      maxWidthClass: "max-w-md",
+    });
+    if (!choice) return Promise.resolve(record);
+
+    const promise = recordRequest
+      .del({}, `/${zone.id}/records/${record.id}`)
+      .then((record) => {
+        mutate("/dns/zones");
+        return Promise.resolve(record);
+      });
+
+    notify({
+      title: `${record.type} Record '${record.name}'`,
+      description: `DNS Record was deleted successfully.`,
+      promise: promise,
+      loadingMessage: "Deleting DNS Record...",
+    });
+
+    return promise;
+  };
+
+  const openZoneModal = (zone?: DNSZone, distributionGroups?: Group[]) => {
+    if (zone) setCurrentZone(zone);
+    if (distributionGroups) setInitialDistributionGroups(distributionGroups);
+    setDnsModal(true);
+  };
+
+  const openRecordModal = (zone: DNSZone, record?: DNSRecord) => {
+    setCurrentZone(zone);
+    if (record) setCurrentRecord(record);
+    setRecordModal(true);
+  };
+
+  const askForRecord = async (zone: DNSZone) => {
+    const choice = await confirm({
+      title: `Add new record to '${zone.name}'?`,
+      description:
+        "Add either an A, AAAA or a CNAME record to control domain name resolution for your network.",
+      confirmText: "Add Record",
+      cancelText: "Later",
+      type: "default",
+      maxWidthClass: "max-w-md",
+    });
+    if (!choice) return;
+    openRecordModal(zone);
+  };
+
+  return (
+    <DNSZonesContext.Provider
+      value={{
+        createZone,
+        updateZone,
+        deleteZone,
+        openZoneModal,
+        openRecordModal,
+        addRecord,
+        updateRecord,
+        deleteRecord,
+        askForRecord,
+      }}
+    >
+      {children}
+      <DNSZoneModal
+        open={dnsModal}
+        onOpenChange={(open) => {
+          setDnsModal(open);
+          if (!open) {
+            setCurrentZone(undefined);
+            setInitialDistributionGroups(undefined);
+          }
+        }}
+        onSuccessAdded={(z) => askForRecord(z)}
+        zone={currentZone}
+        initialDistributionGroups={initialDistributionGroups}
+      />
+      {currentZone && (
+        <DNSRecordModal
+          open={recordModal}
+          onOpenChange={(open) => {
+            setRecordModal(open);
+            if (!open) {
+              setCurrentZone(undefined);
+              setCurrentRecord(undefined);
+            }
+          }}
+          zone={currentZone}
+          record={currentRecord}
+        />
+      )}
+    </DNSZonesContext.Provider>
+  );
+};
+
+export const useDNSZones = () => React.useContext(DNSZonesContext);
diff --git a/src/modules/dns/zones/records/DNSRecordActionCell.tsx b/src/modules/dns/zones/records/DNSRecordActionCell.tsx
new file mode 100644
index 00000000..399206ee
--- /dev/null
+++ b/src/modules/dns/zones/records/DNSRecordActionCell.tsx
@@ -0,0 +1,40 @@
+import Button from "@components/Button";
+import { PenSquare, Trash2 } from "lucide-react";
+import * as React from "react";
+import { usePermissions } from "@/contexts/PermissionsProvider";
+import { DNSRecord } from "@/interfaces/DNS";
+import { useDNSZones } from "@/modules/dns/zones/DNSZonesProvider";
+import { useDNSZone } from "@/modules/dns/zones/records/DNSRecordsTable";
+
+type Props = {
+  record: DNSRecord;
+};
+
+export const DNSRecordActionCell = ({ record }: Props) => {
+  const { permission } = usePermissions();
+  const { deleteRecord, openRecordModal } = useDNSZones();
+  const zone = useDNSZone();
+
+  return (
+    <div className={"flex justify-end pr-4"}>
+      <Button
+        variant={"default-outline"}
+        size={"sm"}
+        onClick={() => openRecordModal(zone, record)}
+        disabled={!permission?.dns?.update}
+      >
+        <PenSquare size={16} />
+        Edit
+      </Button>
+      <Button
+        variant={"danger-outline"}
+        size={"sm"}
+        onClick={() => deleteRecord(zone, record)}
+        disabled={!permission?.dns?.delete}
+      >
+        <Trash2 size={16} />
+        Delete
+      </Button>
+    </div>
+  );
+};
diff --git a/src/modules/dns/zones/records/DNSRecordContentCell.tsx b/src/modules/dns/zones/records/DNSRecordContentCell.tsx
new file mode 100644
index 00000000..7e04974d
--- /dev/null
+++ b/src/modules/dns/zones/records/DNSRecordContentCell.tsx
@@ -0,0 +1,19 @@
+import CopyToClipboardText from "@components/CopyToClipboardText";
+import * as React from "react";
+import { DNSRecord } from "@/interfaces/DNS";
+
+type Props = {
+  record: DNSRecord;
+};
+
+export const DNSRecordContentCell = ({ record }: Props) => {
+  return (
+    <div className="flex flex-col gap-0 dark:text-neutral-300 text-neutral-500 font-light truncate font-mono">
+      <CopyToClipboardText>
+        <span className={"font-normal truncate text-[0.82rem]"}>
+          {record.content}
+        </span>
+      </CopyToClipboardText>
+    </div>
+  );
+};
diff --git a/src/modules/dns/zones/records/DNSRecordNameCell.tsx b/src/modules/dns/zones/records/DNSRecordNameCell.tsx
new file mode 100644
index 00000000..3eed5c78
--- /dev/null
+++ b/src/modules/dns/zones/records/DNSRecordNameCell.tsx
@@ -0,0 +1,17 @@
+import CopyToClipboardText from "@components/CopyToClipboardText";
+import * as React from "react";
+import { DNSRecord } from "@/interfaces/DNS";
+
+type Props = {
+  record: DNSRecord;
+};
+
+export const DNSRecordNameCell = ({ record }: Props) => {
+  return (
+    <div className="flex flex-col gap-0 dark:text-neutral-300 text-neutral-500 font-light truncate">
+      <CopyToClipboardText>
+        <span className={"font-normal truncate"}>{record.name}</span>
+      </CopyToClipboardText>
+    </div>
+  );
+};
diff --git a/src/modules/dns/zones/records/DNSRecordTimeToLiveCell.tsx b/src/modules/dns/zones/records/DNSRecordTimeToLiveCell.tsx
new file mode 100644
index 00000000..0dccd9bf
--- /dev/null
+++ b/src/modules/dns/zones/records/DNSRecordTimeToLiveCell.tsx
@@ -0,0 +1,21 @@
+import { ClockIcon } from "lucide-react";
+import * as React from "react";
+import { DNSRecord } from "@/interfaces/DNS";
+import { getTTLLabel } from "@/modules/dns/zones/DNSRecordModal";
+
+type Props = {
+  record: DNSRecord;
+};
+
+export const DNSRecordTimeToLiveCell = ({ record }: Props) => {
+  return (
+    <div
+      className={
+        "flex items-center whitespace-nowrap gap-2 dark:text-neutral-300 text-neutral-500 hover:text-neutral-100 transition-all py-2 px-3 rounded-md"
+      }
+    >
+      <ClockIcon size={14} />
+      {getTTLLabel(record.ttl)}
+    </div>
+  );
+};
diff --git a/src/modules/dns/zones/records/DNSRecordTypeCell.tsx b/src/modules/dns/zones/records/DNSRecordTypeCell.tsx
new file mode 100644
index 00000000..5b8faaf9
--- /dev/null
+++ b/src/modules/dns/zones/records/DNSRecordTypeCell.tsx
@@ -0,0 +1,20 @@
+import Badge from "@components/Badge";
+import * as React from "react";
+import { DNSRecord } from "@/interfaces/DNS";
+
+type Props = {
+  record: DNSRecord;
+};
+
+export const DNSRecordTypeCell = ({ record }: Props) => {
+  return (
+    <div className={"flex"}>
+      <Badge
+        variant={"gray"}
+        className={"uppercase tracking-wider font-medium"}
+      >
+        {record.type}
+      </Badge>
+    </div>
+  );
+};
diff --git a/src/modules/dns/zones/records/DNSRecordsTable.tsx b/src/modules/dns/zones/records/DNSRecordsTable.tsx
new file mode 100644
index 00000000..ff3fa302
--- /dev/null
+++ b/src/modules/dns/zones/records/DNSRecordsTable.tsx
@@ -0,0 +1,80 @@
+import { DataTable } from "@components/table/DataTable";
+import DataTableHeader from "@components/table/DataTableHeader";
+import { ColumnDef, SortingState } from "@tanstack/react-table";
+import React, { createContext, useContext, useState } from "react";
+import { DNSRecord, DNSZone } from "@/interfaces/DNS";
+import { DNSRecordActionCell } from "@/modules/dns/zones/records/DNSRecordActionCell";
+import { DNSRecordContentCell } from "@/modules/dns/zones/records/DNSRecordContentCell";
+import { DNSRecordNameCell } from "@/modules/dns/zones/records/DNSRecordNameCell";
+import { DNSRecordTimeToLiveCell } from "@/modules/dns/zones/records/DNSRecordTimeToLiveCell";
+import { DNSRecordTypeCell } from "@/modules/dns/zones/records/DNSRecordTypeCell";
+
+type Props = {
+  zone: DNSZone;
+};
+
+export const DNSRecordsTableColumns: ColumnDef<DNSRecord>[] = [
+  {
+    accessorKey: "type",
+    header: ({ column }) => {
+      return <DataTableHeader column={column}>Type</DataTableHeader>;
+    },
+    cell: ({ row }) => <DNSRecordTypeCell record={row.original} />,
+  },
+  {
+    accessorKey: "name",
+    header: ({ column }) => {
+      return <DataTableHeader column={column}>Hostname</DataTableHeader>;
+    },
+    cell: ({ row }) => <DNSRecordNameCell record={row.original} />,
+  },
+  {
+    accessorKey: "content",
+    header: ({ column }) => {
+      return <DataTableHeader column={column}>Content</DataTableHeader>;
+    },
+    cell: ({ row }) => <DNSRecordContentCell record={row.original} />,
+  },
+  {
+    accessorKey: "ttl",
+    header: ({ column }) => {
+      return <DataTableHeader column={column}>TTL</DataTableHeader>;
+    },
+    cell: ({ row }) => <DNSRecordTimeToLiveCell record={row.original} />,
+  },
+  {
+    accessorKey: "id",
+    header: "",
+    cell: ({ row }) => <DNSRecordActionCell record={row.original} />,
+  },
+];
+
+const ZoneContext = createContext({} as DNSZone);
+
+export default function DNSRecordsTable({ zone }: Props) {
+  const [sorting, setSorting] = useState<SortingState>([]);
+
+  return (
+    <ZoneContext.Provider value={zone}>
+      <DataTable
+        uniqueKey={zone.id}
+        keepStateInLocalStorage={false}
+        tableClassName={"mt-0"}
+        minimal={true}
+        showSearchAndFilters={false}
+        rowClassName={"last:pb-10"}
+        className={"bg-nb-gray-960 py-2"}
+        inset={true}
+        text={"DNS Records"}
+        manualPagination={true}
+        sorting={sorting}
+        columnVisibility={{}}
+        setSorting={setSorting}
+        columns={DNSRecordsTableColumns}
+        data={zone.records}
+      />
+    </ZoneContext.Provider>
+  );
+}
+
+export const useDNSZone = () => useContext(ZoneContext);
diff --git a/src/modules/dns/zones/table/DNSZonesActionCell.tsx b/src/modules/dns/zones/table/DNSZonesActionCell.tsx
new file mode 100644
index 00000000..aa1a490d
--- /dev/null
+++ b/src/modules/dns/zones/table/DNSZonesActionCell.tsx
@@ -0,0 +1,58 @@
+import Button from "@components/Button";
+import {
+  DropdownMenu,
+  DropdownMenuContent,
+  DropdownMenuItem,
+  DropdownMenuTrigger,
+} from "@components/DropdownMenu";
+import { MoreVertical, SquarePenIcon, Trash2 } from "lucide-react";
+import * as React from "react";
+import { usePermissions } from "@/contexts/PermissionsProvider";
+import { DNSZone } from "@/interfaces/DNS";
+import { useDNSZones } from "@/modules/dns/zones/DNSZonesProvider";
+
+type Props = {
+  zone: DNSZone;
+};
+
+export const DNSZonesActionCell = ({ zone }: Props) => {
+  const { permission } = usePermissions();
+  const { openZoneModal, deleteZone } = useDNSZones();
+
+  return (
+    <div className={"flex justify-end pr-4"}>
+      <DropdownMenu modal={false}>
+        <DropdownMenuTrigger
+          asChild={true}
+          onClick={(e) => {
+            e.stopPropagation();
+            e.preventDefault();
+          }}
+        >
+          <Button variant={"secondary"} className={"!px-3"}>
+            <MoreVertical size={16} className={"shrink-0"} />
+          </Button>
+        </DropdownMenuTrigger>
+        <DropdownMenuContent className="w-auto" align="end">
+          <DropdownMenuItem onClick={() => openZoneModal(zone)}>
+            <div className={"flex gap-3 items-center"}>
+              <SquarePenIcon size={14} className={"shrink-0"} />
+              Edit
+            </div>
+          </DropdownMenuItem>
+
+          <DropdownMenuItem
+            onClick={() => deleteZone(zone)}
+            variant={"danger"}
+            disabled={!permission?.dns?.delete}
+          >
+            <div className={"flex gap-3 items-center"}>
+              <Trash2 size={14} className={"shrink-0"} />
+              Delete
+            </div>
+          </DropdownMenuItem>
+        </DropdownMenuContent>
+      </DropdownMenu>
+    </div>
+  );
+};
diff --git a/src/modules/dns/zones/table/DNSZonesActiveCell.tsx b/src/modules/dns/zones/table/DNSZonesActiveCell.tsx
new file mode 100644
index 00000000..35daa15e
--- /dev/null
+++ b/src/modules/dns/zones/table/DNSZonesActiveCell.tsx
@@ -0,0 +1,32 @@
+import { ToggleSwitch } from "@components/ToggleSwitch";
+import * as React from "react";
+import { usePermissions } from "@/contexts/PermissionsProvider";
+import { DNSZone } from "@/interfaces/DNS";
+import { useDNSZones } from "@/modules/dns/zones/DNSZonesProvider";
+
+type Props = {
+  zone: DNSZone;
+};
+
+export const DNSZonesActiveCell = ({ zone }: Props) => {
+  const { permission } = usePermissions();
+  const { updateZone } = useDNSZones();
+
+  return (
+    <div className={"flex min-w-[0px]"}>
+      <ToggleSwitch
+        disabled={!permission?.dns?.update}
+        checked={zone.enabled}
+        size={"small"}
+        onClick={(e) => {
+          e.preventDefault();
+          e.stopPropagation();
+          updateZone({
+            ...zone,
+            enabled: !zone.enabled,
+          });
+        }}
+      />
+    </div>
+  );
+};
diff --git a/src/modules/dns/zones/table/DNSZonesGroupCell.tsx b/src/modules/dns/zones/table/DNSZonesGroupCell.tsx
new file mode 100644
index 00000000..3b3fa487
--- /dev/null
+++ b/src/modules/dns/zones/table/DNSZonesGroupCell.tsx
@@ -0,0 +1,60 @@
+import * as React from "react";
+import { useMemo, useState } from "react";
+import { useGroups } from "@/contexts/GroupsProvider";
+import { usePermissions } from "@/contexts/PermissionsProvider";
+import { DNSZone } from "@/interfaces/DNS";
+import { Group } from "@/interfaces/Group";
+import EmptyRow from "@/modules/common-table-rows/EmptyRow";
+import GroupsRow from "@/modules/common-table-rows/GroupsRow";
+import { useDNSZones } from "@/modules/dns/zones/DNSZonesProvider";
+
+type Props = {
+  zone: DNSZone;
+};
+
+export const DNSZonesGroupCell = ({ zone }: Props) => {
+  const { groups } = useGroups();
+  const { updateZone } = useDNSZones();
+  const [modal, setModal] = useState(false);
+  const { permission } = usePermissions();
+
+  const allGroups = zone?.distribution_groups
+    .map((group) => {
+      return groups?.find((g) => g.id == group);
+    })
+    .filter((g) => g != undefined) as Group[];
+
+  const groupIDs = useMemo(() => {
+    return allGroups
+      ?.map((group) => group.id)
+      .filter((id) => id !== undefined) as string[];
+  }, [allGroups]);
+
+  const handleSave = async (promises: Promise<Group>[]) => {
+    const groups = await Promise.all(promises);
+    const groupIds = groups?.map((g) => g.id as string);
+    await updateZone({
+      ...zone,
+      distribution_groups: groupIds,
+    }).then(() => {
+      setModal(false);
+    });
+  };
+
+  if (!zone?.distribution_groups) return <EmptyRow />;
+
+  return (
+    <GroupsRow
+      label={"Distribution Groups"}
+      description={
+        "Advertise this zone to peers that belong to the following groups"
+      }
+      groups={groupIDs || []}
+      hideAllGroup={false}
+      disabled={!permission?.dns?.update}
+      onSave={handleSave}
+      modal={modal}
+      setModal={setModal}
+    />
+  );
+};
diff --git a/src/modules/dns/zones/table/DNSZonesNameCell.tsx b/src/modules/dns/zones/table/DNSZonesNameCell.tsx
new file mode 100644
index 00000000..d534c03d
--- /dev/null
+++ b/src/modules/dns/zones/table/DNSZonesNameCell.tsx
@@ -0,0 +1,38 @@
+import { cn } from "@utils/helpers";
+import { ChevronDown, ChevronRightIcon } from "lucide-react";
+import * as React from "react";
+import { DNSZone } from "@/interfaces/DNS";
+import ActiveInactiveRow from "@/modules/common-table-rows/ActiveInactiveRow";
+
+type Props = {
+  zone: DNSZone;
+};
+
+export const DNSZonesNameCell = ({ zone }: Props) => {
+  const hasRecords = (zone?.records?.length ?? 0) > 0;
+
+  return (
+    <div className={"flex gap-6 items-center min-w-[270px] max-w-[270px]"}>
+      <ChevronRightIcon
+        size={20}
+        className={cn(
+          "group-data-[accordion=opened]/accordion:hidden text-nb-gray-400 shrink-0",
+          !hasRecords && "cursor-default opacity-0",
+        )}
+      />
+      <ChevronDown
+        size={20}
+        className={cn(
+          "group-data-[accordion=closed]/accordion:hidden text-nb-gray-400 shrink-0",
+          !hasRecords && "cursor-default opacity-0",
+        )}
+      />
+      <ActiveInactiveRow
+        active={zone.enabled}
+        inactiveDot={"gray"}
+        text={zone.domain}
+        dataCy={zone.id}
+      />
+    </div>
+  );
+};
diff --git a/src/modules/dns/zones/table/DNSZonesRecordsCell.tsx b/src/modules/dns/zones/table/DNSZonesRecordsCell.tsx
new file mode 100644
index 00000000..2607f428
--- /dev/null
+++ b/src/modules/dns/zones/table/DNSZonesRecordsCell.tsx
@@ -0,0 +1,47 @@
+import Badge from "@components/Badge";
+import Button from "@components/Button";
+import { GlobeIcon, PlusCircle } from "lucide-react";
+import * as React from "react";
+import { usePermissions } from "@/contexts/PermissionsProvider";
+import { DNSZone } from "@/interfaces/DNS";
+import { useDNSZones } from "@/modules/dns/zones/DNSZonesProvider";
+
+type Props = {
+  zone: DNSZone;
+};
+
+export const DNSZonesRecordsCell = ({ zone }: Props) => {
+  const { permission } = usePermissions();
+  const { openRecordModal } = useDNSZones();
+
+  const recordsCount = zone?.records?.length ?? 0;
+
+  return (
+    <div className={"flex gap-3"}>
+      {recordsCount > 0 && (
+        <Badge
+          variant={"gray"}
+          useHover={true}
+          className={"cursor-pointer"}
+          onClick={() => void 0}
+        >
+          <GlobeIcon size={12} />
+          <div>
+            <span className={"font-medium text-xs"}>{recordsCount}</span>
+          </div>
+        </Badge>
+      )}
+
+      <Button
+        size={"xs"}
+        variant={"secondary"}
+        className={"min-w-[130px]"}
+        onClick={() => openRecordModal(zone)}
+        disabled={!permission?.dns?.create}
+      >
+        <PlusCircle size={12} />
+        Add Record
+      </Button>
+    </div>
+  );
+};
diff --git a/src/modules/dns/zones/table/DNSZonesSearchDomainCell.tsx b/src/modules/dns/zones/table/DNSZonesSearchDomainCell.tsx
new file mode 100644
index 00000000..2fd613d4
--- /dev/null
+++ b/src/modules/dns/zones/table/DNSZonesSearchDomainCell.tsx
@@ -0,0 +1,32 @@
+import { ToggleSwitch } from "@components/ToggleSwitch";
+import * as React from "react";
+import { usePermissions } from "@/contexts/PermissionsProvider";
+import { DNSZone } from "@/interfaces/DNS";
+import { useDNSZones } from "@/modules/dns/zones/DNSZonesProvider";
+
+type Props = {
+  zone: DNSZone;
+};
+
+export const DNSZonesSearchDomainCell = ({ zone }: Props) => {
+  const { permission } = usePermissions();
+  const { updateZone } = useDNSZones();
+
+  return (
+    <div className={"flex min-w-[0px]"}>
+      <ToggleSwitch
+        disabled={!permission?.dns?.update}
+        checked={zone?.enable_search_domain}
+        size={"small"}
+        onClick={(e) => {
+          e.preventDefault();
+          e.stopPropagation();
+          updateZone({
+            ...zone,
+            enable_search_domain: !zone.enable_search_domain,
+          });
+        }}
+      />
+    </div>
+  );
+};
diff --git a/src/modules/dns/zones/table/DNSZonesTable.tsx b/src/modules/dns/zones/table/DNSZonesTable.tsx
new file mode 100644
index 00000000..ff7e946f
--- /dev/null
+++ b/src/modules/dns/zones/table/DNSZonesTable.tsx
@@ -0,0 +1,303 @@
+import Button from "@components/Button";
+import ButtonGroup from "@components/ButtonGroup";
+import Card from "@components/Card";
+import InlineLink from "@components/InlineLink";
+import SquareIcon from "@components/SquareIcon";
+import { DataTable } from "@components/table/DataTable";
+import DataTableHeader from "@components/table/DataTableHeader";
+import DataTableRefreshButton from "@components/table/DataTableRefreshButton";
+import { DataTableRowsPerPage } from "@components/table/DataTableRowsPerPage";
+import GetStartedTest from "@components/ui/GetStartedTest";
+import NoResults from "@components/ui/NoResults";
+import { ColumnDef, SortingState } from "@tanstack/react-table";
+import { ExternalLinkIcon, PlusCircle } from "lucide-react";
+import { usePathname } from "next/navigation";
+import React, { useMemo } from "react";
+import { useSWRConfig } from "swr";
+import { usePermissions } from "@/contexts/PermissionsProvider";
+import { useLocalStorage } from "@/hooks/useLocalStorage";
+import { DNS_ZONE_DOCS_LINK, DNSZone } from "@/interfaces/DNS";
+import { useDNSZones } from "@/modules/dns/zones/DNSZonesProvider";
+import DNSRecordsTable from "@/modules/dns/zones/records/DNSRecordsTable";
+import { DNSZonesActionCell } from "@/modules/dns/zones/table/DNSZonesActionCell";
+import { DNSZonesActiveCell } from "@/modules/dns/zones/table/DNSZonesActiveCell";
+import { DNSZonesGroupCell } from "@/modules/dns/zones/table/DNSZonesGroupCell";
+import { DNSZonesNameCell } from "@/modules/dns/zones/table/DNSZonesNameCell";
+import { DNSZonesRecordsCell } from "@/modules/dns/zones/table/DNSZonesRecordsCell";
+import { DNSZonesSearchDomainCell } from "@/modules/dns/zones/table/DNSZonesSearchDomainCell";
+import { Group } from "@/interfaces/Group";
+import DNSZoneIcon from "@/assets/icons/DNSZoneIcon";
+import { useGroups } from "@/contexts/GroupsProvider";
+
+export const DNSZonesColumns: ColumnDef<DNSZone>[] = [
+  {
+    accessorKey: "domain",
+    header: ({ column }) => (
+      <DataTableHeader column={column}>Zone</DataTableHeader>
+    ),
+    sortingFn: "text",
+    cell: ({ row }) => <DNSZonesNameCell zone={row.original} />,
+  },
+  {
+    accessorKey: "enabled",
+    header: ({ column }) => (
+      <DataTableHeader column={column}>Active</DataTableHeader>
+    ),
+    cell: ({ row }) => <DNSZonesActiveCell zone={row.original} />,
+  },
+  {
+    accessorKey: "records",
+    header: ({ column }) => (
+      <DataTableHeader column={column}>Records</DataTableHeader>
+    ),
+    sortingFn: "text",
+    cell: ({ row }) => <DNSZonesRecordsCell zone={row.original} />,
+  },
+  {
+    accessorKey: "distribution_groups",
+    header: ({ column }) => (
+      <DataTableHeader column={column}>Distribution Groups</DataTableHeader>
+    ),
+    cell: ({ row }) => <DNSZonesGroupCell zone={row.original} />,
+  },
+  {
+    accessorKey: "enable_search_domain",
+    header: ({ column }) => (
+      <DataTableHeader column={column}>Search Domain</DataTableHeader>
+    ),
+    cell: ({ row }) => <DNSZonesSearchDomainCell zone={row.original} />,
+  },
+  {
+    accessorKey: "id",
+    header: () => "",
+    cell: ({ row }) => <DNSZonesActionCell zone={row.original} />,
+  },
+  {
+    id: "searchString",
+    accessorFn: (row) => {
+      return [
+        row?.groups_search,
+        row?.name,
+        row?.domain,
+        row?.records?.map((r) => r.name).join(""),
+        row?.records?.map((r) => r.content).join(""),
+        row?.records?.map((r) => r.type).join(""),
+      ]?.join("");
+    },
+  },
+];
+
+type Props = {
+  isLoading: boolean;
+  data?: DNSZone[];
+  headingTarget?: HTMLHeadingElement | null;
+  isGroupPage?: boolean;
+  distributionGroups?: Group[];
+};
+
+export default function DNSZonesTable({
+  data,
+  isLoading,
+  headingTarget,
+  isGroupPage = false,
+  distributionGroups,
+}: Props) {
+  const { mutate } = useSWRConfig();
+  const path = usePathname();
+  const { groups } = useGroups();
+
+  // Default sorting state of the table
+  const [sorting, setSorting] = useLocalStorage<SortingState>(
+    "netbird-table-sort" + path,
+    [
+      {
+        id: "domain",
+        desc: true,
+      },
+      {
+        id: "id",
+        desc: true,
+      },
+    ],
+    !isGroupPage,
+  );
+
+  const zonesWithGroups = useMemo(() => {
+    return (
+      data?.map((zone) => {
+        return {
+          ...zone,
+          groups_search: groups
+            ?.map((g) =>
+              zone?.distribution_groups?.includes(g?.id ?? "") ? g.name : "",
+            )
+            .join(""),
+        } as DNSZone;
+      }) ?? []
+    );
+  }, [data, groups]);
+
+  return (
+    <DataTable
+      headingTarget={headingTarget}
+      isLoading={isLoading}
+      text={"DNS Zones"}
+      sorting={sorting}
+      setSorting={setSorting}
+      columns={DNSZonesColumns}
+      data={zonesWithGroups}
+      useRowId={true}
+      wrapperComponent={isGroupPage ? Card : undefined}
+      wrapperProps={isGroupPage ? { className: "mt-6 w-full" } : undefined}
+      paginationPaddingClassName={isGroupPage ? "px-0 pt-8" : undefined}
+      tableClassName={isGroupPage ? "mt-0 mb-2" : undefined}
+      inset={false}
+      minimal={isGroupPage}
+      keepStateInLocalStorage={!isGroupPage}
+      searchPlaceholder={"Search by domain, ip, content or group..."}
+      columnVisibility={{ searchString: false }}
+      renderExpandedRow={(zone) => {
+        const hasRecords = (zone?.records?.length ?? 0) > 0;
+        if (!hasRecords) return;
+        return (
+          <>
+            <DNSRecordsTable zone={zone} />
+            <div className={"h-2 w-full bg-nb-gray-960"}></div>
+          </>
+        );
+      }}
+      getStartedCard={
+        isGroupPage ? (
+          <NoResults
+            icon={<DNSZoneIcon className={"fill-nb-gray-200"} size={24} />}
+            className={"py-4"}
+            contentClassName={"max-w-lg"}
+            title={"This group is not used within any zones yet"}
+            description={
+              "Assign this group as a distribution group in your zones to see them listed here."
+            }
+          >
+            <div className={"gap-x-4 flex items-center justify-center mt-4"}>
+              <AddZoneButton distributionGroups={distributionGroups} />
+            </div>
+          </NoResults>
+        ) : (
+          <GetStartedTest
+            icon={
+              <SquareIcon
+                icon={<DNSZoneIcon className={"fill-nb-gray-200"} size={24} />}
+                color={"gray"}
+                size={"large"}
+              />
+            }
+            title={"Create New Zone"}
+            description={
+              "It looks like you don't have any zones. Control domain name resolution for your network by adding a zone."
+            }
+            button={
+              <div className={"gap-x-4 flex items-center justify-center"}>
+                <AddZoneButton distributionGroups={distributionGroups} />
+              </div>
+            }
+            learnMore={
+              <>
+                Learn more about
+                <InlineLink href={DNS_ZONE_DOCS_LINK} target={"_blank"}>
+                  DNS Zones
+                  <ExternalLinkIcon size={12} />
+                </InlineLink>
+              </>
+            }
+          />
+        )
+      }
+      rightSide={() => (
+        <>
+          {data && data?.length > 0 && (
+            <div className={"gap-x-4 ml-auto flex"}>
+              <AddZoneButton distributionGroups={distributionGroups} />
+            </div>
+          )}
+        </>
+      )}
+    >
+      {(table) => (
+        <>
+          <ButtonGroup disabled={data?.length == 0}>
+            <ButtonGroup.Button
+              onClick={() => {
+                table.setPageIndex(0);
+                table.getColumn("enabled")?.setFilterValue(undefined);
+              }}
+              disabled={data?.length == 0}
+              variant={
+                table.getColumn("enabled")?.getFilterValue() === undefined
+                  ? "tertiary"
+                  : "secondary"
+              }
+            >
+              All
+            </ButtonGroup.Button>
+            <ButtonGroup.Button
+              onClick={() => {
+                table.setPageIndex(0);
+                table.getColumn("enabled")?.setFilterValue(true);
+              }}
+              disabled={data?.length == 0}
+              variant={
+                table.getColumn("enabled")?.getFilterValue() === true
+                  ? "tertiary"
+                  : "secondary"
+              }
+            >
+              Active
+            </ButtonGroup.Button>
+            <ButtonGroup.Button
+              onClick={() => {
+                table.setPageIndex(0);
+                table.getColumn("enabled")?.setFilterValue(false);
+              }}
+              disabled={data?.length == 0}
+              variant={
+                table.getColumn("enabled")?.getFilterValue() === false
+                  ? "tertiary"
+                  : "secondary"
+              }
+            >
+              Inactive
+            </ButtonGroup.Button>
+          </ButtonGroup>
+          <DataTableRowsPerPage table={table} disabled={data?.length == 0} />
+          <DataTableRefreshButton
+            isDisabled={data?.length == 0}
+            onClick={() => {
+              mutate("/dns/zones").then();
+              mutate("/groups").then();
+            }}
+          />
+        </>
+      )}
+    </DataTable>
+  );
+}
+
+type AddZoneButtonProps = {
+  distributionGroups?: Group[];
+};
+
+const AddZoneButton = ({ distributionGroups }: AddZoneButtonProps) => {
+  const { permission } = usePermissions();
+  const { openZoneModal } = useDNSZones();
+
+  return (
+    <Button
+      variant={"primary"}
+      className={""}
+      disabled={!permission?.dns?.create}
+      onClick={() => openZoneModal(undefined, distributionGroups)}
+    >
+      <PlusCircle size={16} />
+      Add Zone
+    </Button>
+  );
+};
diff --git a/src/modules/groups/details/GroupDNSZonesSection.tsx b/src/modules/groups/details/GroupDNSZonesSection.tsx
new file mode 100644
index 00000000..fecac435
--- /dev/null
+++ b/src/modules/groups/details/GroupDNSZonesSection.tsx
@@ -0,0 +1,29 @@
+import React from "react";
+import { useGroupContext } from "@/contexts/GroupProvider";
+import { DNSZone } from "@/interfaces/DNS";
+import { DNSZonesProvider } from "@/modules/dns/zones/DNSZonesProvider";
+import DNSZonesTable from "@/modules/dns/zones/table/DNSZonesTable";
+import { GroupDetailsTableContainer } from "@/modules/groups/details/GroupDetailsTableContainer";
+
+export const GroupDNSZonesSection = ({
+  zones,
+  isLoading = true,
+}: {
+  zones?: DNSZone[];
+  isLoading?: boolean;
+}) => {
+  const { group } = useGroupContext();
+
+  return (
+    <GroupDetailsTableContainer>
+      <DNSZonesProvider>
+        <DNSZonesTable
+          isGroupPage={true}
+          isLoading={isLoading}
+          data={zones}
+          distributionGroups={[group]}
+        />
+      </DNSZonesProvider>
+    </GroupDetailsTableContainer>
+  );
+};
diff --git a/src/modules/groups/details/GroupNameserversSection.tsx b/src/modules/groups/details/GroupNameserversSection.tsx
index f4bd7014..15a1fa85 100644
--- a/src/modules/groups/details/GroupNameserversSection.tsx
+++ b/src/modules/groups/details/GroupNameserversSection.tsx
@@ -4,7 +4,7 @@ import { NameserverGroup } from "@/interfaces/Nameserver";
 import { GroupDetailsTableContainer } from "@/modules/groups/details/GroupDetailsTableContainer";
 
 const NameserverGroupTable = lazy(
-  () => import("@/modules/dns-nameservers/table/NameserverGroupTable"),
+  () => import("@/modules/dns/nameservers/table/NameserverGroupTable"),
 );
 
 type Props = {
diff --git a/src/modules/groups/details/useGroupDetails.ts b/src/modules/groups/details/useGroupDetails.ts
index d16ea66d..85e07777 100644
--- a/src/modules/groups/details/useGroupDetails.ts
+++ b/src/modules/groups/details/useGroupDetails.ts
@@ -1,4 +1,5 @@
 import { useMemo } from "react";
+import { DNSZone } from "@/interfaces/DNS";
 import { Group, GroupPeer, GroupResource } from "@/interfaces/Group";
 import { NameserverGroup } from "@/interfaces/Nameserver";
 import {
@@ -16,6 +17,7 @@ import useFetchApi from "@/utils/api";
 export interface GroupDetails extends Group {
   policies: Policy[];
   nameservers: NameserverGroup[];
+  zones?: DNSZone[];
   routes: Route[];
   setupKeys: SetupKey[];
   users: User[];
@@ -31,6 +33,8 @@ export default function useGroupDetails(groupId: string) {
     useFetchApi<Policy[]>(`/policies`);
   const { data: nameservers, isLoading: isNameserversLoading } =
     useFetchApi<NameserverGroup[]>(`/dns/nameservers`);
+  const { data: zones, isLoading: isZonesLoading } =
+    useFetchApi<DNSZone[]>(`/dns/zones`);
   const { data: routes, isLoading: isRoutesLoading } =
     useFetchApi<Route[]>(`/routes`);
   const { data: setupKeys, isLoading: isSetupKeysLoading } =
@@ -65,6 +69,12 @@ export default function useGroupDetails(groupId: string) {
     return nameservers?.filter((ns) => ns.groups?.includes(groupId)) || [];
   }, [nameservers, groupId]);
 
+  const linkedZones = useMemo(() => {
+    return (
+      zones?.filter((ns) => ns.distribution_groups?.includes(groupId)) || []
+    );
+  }, [zones, groupId]);
+
   const linkedRoutes = useMemo(() => {
     return (
       routes?.filter((route) => {
@@ -117,6 +127,7 @@ export default function useGroupDetails(groupId: string) {
     isGroupsLoading ||
     isPoliciesLoading ||
     isNameserversLoading ||
+    isZonesLoading ||
     isRoutesLoading ||
     isSetupKeysLoading ||
     isUsersLoading ||
@@ -130,6 +141,7 @@ export default function useGroupDetails(groupId: string) {
       ...group,
       policies: linkedPolicies,
       nameservers: linkedNameservers,
+      zones: linkedZones,
       routes: linkedRoutes,
       setupKeys: linkedSetupKeys,
       users: linkedUsers,
@@ -141,6 +153,7 @@ export default function useGroupDetails(groupId: string) {
     group,
     linkedPolicies,
     linkedNameservers,
+    linkedZones,
     linkedRoutes,
     linkedSetupKeys,
     linkedUsers,
diff --git a/src/modules/groups/table/GroupsTable.tsx b/src/modules/groups/table/GroupsTable.tsx
index e0894bfe..67addd2a 100644
--- a/src/modules/groups/table/GroupsTable.tsx
+++ b/src/modules/groups/table/GroupsTable.tsx
@@ -3,6 +3,7 @@ import { DataTable } from "@components/table/DataTable";
 import DataTableHeader from "@components/table/DataTableHeader";
 import { DataTableRowsPerPage } from "@components/table/DataTableRowsPerPage";
 import { ColumnDef, SortingState } from "@tanstack/react-table";
+import { removeAllSpaces } from "@utils/helpers";
 import { Layers3Icon } from "lucide-react";
 import { usePathname } from "next/navigation";
 import React from "react";
@@ -19,7 +20,7 @@ import GroupsActionCell from "@/modules/groups/table/GroupsActionCell";
 import GroupsCountCell from "@/modules/groups/table/GroupsCountCell";
 import GroupsNameCell from "@/modules/groups/table/GroupsNameCell";
 import useGroupsUsage, { GroupUsage } from "@/modules/groups/useGroupsUsage";
-import { removeAllSpaces } from "@utils/helpers";
+import DNSZoneIcon from "@/assets/icons/DNSZoneIcon";
 
 export const GroupsTableColumns: ColumnDef<GroupUsage>[] = [
   {
@@ -178,6 +179,28 @@ export const GroupsTableColumns: ColumnDef<GroupUsage>[] = [
       />
     ),
   },
+  {
+    accessorKey: "zones_count",
+    header: ({ column }) => {
+      return (
+        <DataTableHeader
+          column={column}
+          tooltip={<div className={"text-xs normal-case"}>Zones</div>}
+        >
+          <DNSZoneIcon size={16} />
+        </DataTableHeader>
+      );
+    },
+    cell: ({ row }) => (
+      <GroupsCountCell
+        icon={<DNSZoneIcon size={14} />}
+        groupName={row.original.name}
+        href={`/group?id=${row.original.id}&tab=zones`}
+        text={"Zone(s)"}
+        count={row.original.zones_count}
+      />
+    ),
+  },
   {
     accessorKey: "setup_keys_count",
     header: ({ column }) => {
@@ -216,7 +239,8 @@ export const GroupsTableColumns: ColumnDef<GroupUsage>[] = [
         row.routes_count > 0 ||
         row.setup_keys_count > 0 ||
         row.users_count > 0 ||
-        row.resources_count > 0
+        row.resources_count > 0 ||
+        row.zones_count
       );
     },
   },
diff --git a/src/modules/groups/useGroupsUsage.tsx b/src/modules/groups/useGroupsUsage.tsx
index 22fef68d..1fb5e5bc 100644
--- a/src/modules/groups/useGroupsUsage.tsx
+++ b/src/modules/groups/useGroupsUsage.tsx
@@ -1,5 +1,6 @@
 import useFetchApi from "@utils/api";
 import { useMemo } from "react";
+import { DNSZone } from "@/interfaces/DNS";
 import { Group } from "@/interfaces/Group";
 import { NameserverGroup } from "@/interfaces/Nameserver";
 import { Policy } from "@/interfaces/Policy";
@@ -11,6 +12,7 @@ export interface GroupUsage extends Group {
   peers_count: number;
   policies_count: number;
   nameservers_count: number;
+  zones_count: number;
   routes_count: number;
   setup_keys_count: number;
   users_count: number;
@@ -24,6 +26,8 @@ export default function useGroupsUsage() {
     useFetchApi<Policy[]>(`/policies`); // Policies
   const { data: nameservers, isLoading: isNameserversLoading } =
     useFetchApi<NameserverGroup[]>(`/dns/nameservers`); // DNS
+  const { data: zones, isLoading: isZonesLoading } =
+    useFetchApi<DNSZone[]>(`/dns/zones`); // DNS Zones
   const { data: routes, isLoading: isRoutesLoading } =
     useFetchApi<Route[]>(`/routes`); // Routes
   const { data: setupKeys, isLoading: isSetupKeysLoading } =
@@ -57,6 +61,14 @@ export default function useGroupsUsage() {
       .filter((u) => u !== undefined);
   }, [nameservers, isNameserversLoading]);
 
+  const zonesGroups = useMemo(() => {
+    if (isZonesLoading) return;
+    if (!zones) return [];
+    return zones
+      ?.map((zone) => zone.distribution_groups)
+      .filter((u) => u !== undefined);
+  }, [zones, isZonesLoading]);
+
   const setupKeysGroups = useMemo(() => {
     if (isSetupKeysLoading) return;
     if (!setupKeys) return [];
@@ -78,6 +90,7 @@ export default function useGroupsUsage() {
       isGroupsLoading ||
       isPoliciesLoading ||
       isNameserversLoading ||
+      isZonesLoading ||
       isRoutesLoading ||
       isSetupKeysLoading ||
       isUsersLoading
@@ -86,6 +99,7 @@ export default function useGroupsUsage() {
     isGroupsLoading,
     isPoliciesLoading,
     isNameserversLoading,
+    isZonesLoading,
     isRoutesLoading,
     isSetupKeysLoading,
     isUsersLoading,
@@ -104,6 +118,10 @@ export default function useGroupsUsage() {
         return nameserver.includes(group.id as string);
       }).length;
 
+      const zonesCount = zonesGroups?.filter((zone) => {
+        return zone.includes(group.id as string);
+      }).length;
+
       const routeCount = (
         routes?.filter((route) => {
           const groupId = group.id as string;
@@ -133,6 +151,7 @@ export default function useGroupsUsage() {
         resources_count: group.resources_count,
         policies_count: policyCount,
         nameservers_count: nameserverCount,
+        zones_count: zonesCount,
         routes_count: routeCount,
         setup_keys_count: setupKeyCount,
         users_count: userCount,
@@ -143,6 +162,7 @@ export default function useGroupsUsage() {
     groups,
     policiesGroups,
     nameserversGroups,
+    zonesGroups,
     routes,
     isRoutesLoading,
     setupKeysGroups,
diff --git a/src/modules/routes/RouteTable.tsx b/src/modules/routes/RouteTable.tsx
index 64e941c8..1d628b40 100644
--- a/src/modules/routes/RouteTable.tsx
+++ b/src/modules/routes/RouteTable.tsx
@@ -114,7 +114,7 @@ export default function RouteTable({ row }: Props) {
       desc: true,
     },
   ]);
-  
+
   const hasAtLeastOneExitNode = useMemo(() => {
     return row.routes?.some((route) => route.network === "0.0.0.0/0");
   }, [row.routes]);
@@ -147,7 +147,7 @@ export default function RouteTable({ row }: Props) {
         tableClassName={"mt-0"}
         minimal={true}
         showSearchAndFilters={false}
-        className={"bg-neutral-900/50 py-2"}
+        className={"bg-nb-gray-960 py-2"}
         inset={true}
         text={"Network Routes"}
         manualPagination={true}
diff --git a/tailwind.config.ts b/tailwind.config.ts
index a4321202..58e21f41 100644
--- a/tailwind.config.ts
+++ b/tailwind.config.ts
@@ -29,8 +29,9 @@ const config: Config = {
           "925": "#1e2123",
           "930": "#25282c",
           "935": "#1f2124",
-          "940": "#1c1d21",
+          "940": "#1c1e21",
           "950": "#181a1d",
+          "960": "#15171a",
         },
         netbird: {
           DEFAULT: "#f68330",
